src/basic/util.h

```diff
@@ -174,12 +174,26 @@ static inline void *mempset(void *s, int c, size_t n) {
 }
 
 static inline void _reset_errno_(int *saved_errno) {
-        errno = *saved_errno;
+        if (*saved_errno >= 0)
+                errno = *saved_errno;
 }
 
 #define PROTECT_ERRNO                                                   \
         _cleanup_(_reset_errno_) _unused_ int _saved_errno_ = errno
 
+/*
+ * NSS modules should indicate errors by assigning to the passed-in *errnop
+ * rather than errno directly; however in dynamically-linked programs, errnop
+ * == &errno, so PROTECT_ERRNO has to be disabled in order for assigning to
+ * *errnop to be effective.
+ */
+#define DISARM_PROTECT_ERRNO(r) \
+        ({ \
+                errno = _saved_errno_; \
```

Comment:
please make this idempotent, i.e. if _saved_errno_ is already negative don't restore errno here either